design.md

```diff
@@ -0,0 +1,109 @@
+# Teleport assessment design document
+
+## Goal
+Implement a prototype job worker service that provides an API to run arbitrary
+Linux processes.
+
+## Architecture
+The system is made up of 3 components. The worker library, the grpc api and the client/cli interface to run commands
+* The worker library provides methods to start, stop and get the status of a running job.
+A method is also provided to stream the output of a job. The api can also be queried for all previous output of a running job. The server also provides a stream for clients to listen to job output.
+* The grpc api is a grpc service that sits on top of the worker library. It provides methods to start, stop and get the status of a running job remotely. 
+* The client/cli is an interface for users to start, stop and get the status of a running job on a remote server.
+
+## Worker library API
+The worker library consists of 2 components. The supervisor and the process.
+### Supervisor API
+The supervisor is responsible for managing all jobs in the system. It provides apis to start, stop, get the output and get the status of a job. 
+* `Newsupervisor()` - Returns a new instance of a supervisor
+* `Start(cmd string, args []string, opts...Options)` - Starts a process and return a unique job id. The job id is generated by the supervisor.
```

Comment:
how exactly will the ID be generated?
is it a randomized string or something else?

Reply:
The ID will be a randomized string using a uuid package